### PR TITLE
Use random ports for pre-packaged Tor daemon

### PR DESCRIPTION
**Pull request by rorp**

### CODE REVIEW

File: tor/src/main/scala/org/bitcoins/tor/config/TorAppConfig.scala [hunk 1]
```diff
@@ -181,4 +198,27 @@ object TorAppConfig extends AppConfigFactory[TorAppConfig] {
...
       ec: ExecutionContext): TorAppConfig =
     TorAppConfig(datadir, confs: _*)
+
+  lazy val randomSocks5Port: Int = ports.proxyPort
+
+  lazy val randomControlPort: Int = ports.controlPort
+
+  private case class TorPorts(proxyPort: Int, controlPort: Int)
+
+  private lazy val ports = {
```
Comment (benthecarman) [inline tor/src/main/scala/org/bitcoins/tor/config/TorAppConfig.scala:220]: should probably make these `def`s so if we spin up multiple tor instances they don't overlap

Comment (rorp) [reply tor/src/main/scala/org/bitcoins/tor/config/TorAppConfig.scala:220]: Let's do it when we need multiple instances.

File: db-commons/src/main/resources/reference.conf [hunk 2]
```diff
@@ -61,6 +61,17 @@ bitcoin-s {
...
+    # Tor daemon can be provided by the node operator.
+    # If this parameter set to true, bitcoin-s will connect the provided Tor daemon.
+    # Otherwise bitcoin-s will start its own pre-packaged daemon.
+    provided = false
+
+    # This parameter allows to use random port numbers for pre-packaged Tor daemon,
+    # which is useful if another Tor daemon instance already bound SOCKS5 and control ports.
+    # In this case bitcoin-s.tor.control and bitcoin-s.proxy.socks5
+    # addresses will be automatically changed to "localhost:<random port>"
+    use-random-ports = true
```
Comment (Christewart) [inline db-commons/src/main/resources/reference.conf:73]: So when we ship our binaries, by default we should have this on i think? 

https://github.com/bitcoin-s/bitcoin-s/blob/4ef1e6bcb617c161818b58fcffc470fac3e28c4c/app/bundle/src/main/resources/application.conf#L14

File: testkit/src/main/scala/org/bitcoins/testkit/BitcoinSTestAppConfig.scala [hunk 3]
```diff
@@ -49,6 +49,7 @@ object BitcoinSTestAppConfig {
          |
          |  proxy.enabled = $torEnabled
          |  tor.enabled = $torEnabled
+         |  tor.use-random-ports = false
```
Comment (Christewart) [inline testkit/src/main/scala/org/bitcoins/testkit/BitcoinSTestAppConfig.scala:52]: Shouldn't we be using random ports in tests?

File: db-commons/src/main/resources/reference.conf (hunk 2 above)
Comment (rorp) [reply db-commons/src/main/resources/reference.conf:73]: It's already enabled in `reference.conf`

File: testkit/src/main/scala/org/bitcoins/testkit/BitcoinSTestAppConfig.scala (hunk 3 above)
Comment (rorp) [reply testkit/src/main/scala/org/bitcoins/testkit/BitcoinSTestAppConfig.scala:52]: We could, but the tests running in parallel wouldn't be able reuse a single tor instance.

`Socks5ClientTransportSpec` runs with `tor.use-random-ports = true` thought.

**Review by Christewart — CHANGES_REQUESTED** [review]
This does not seem to work with `bundle/run` on either neutrino backends or bitcoind backends on e1d9fa8

Here is my stack trace for an attempt to connect on neutrino via tor

>[info] Aug 26 08:08:33.357 [notice] Opened Control listener connection (ready) on 127.0.0.1:17137
[info] 2021-08-26T13:09:33UTC ERROR [BitcoinSServerMain] Failed to startup server!
[info] org.bitcoins.asyncutil.AsyncUtil$RpcRetryException: Condition timed out after 60 attempts with interval=1 second waiting periods
[info]  at org.bitcoins.tor.config.TorAppConfig.isBinaryFullyStarted(TorAppConfig.scala:166)
[info]  at org.bitcoins.tor.config.TorAppConfig.$anonfun$start$2(TorAppConfig.scala:129)
[info]  at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:434)
[info]  at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:63)
[info]  at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:100)
[info]  at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[info]  at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:94)
[info]  at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:100)
[info]  at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:49)
[info]  at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:48)
[info]  at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
[info]  at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1016)
[info]  at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1665)
[info]  at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1598)
[info]  at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
[info] 2021-08-26T13:09:33UTC ERROR [BitcoinSServerMain] Failed to initialize configuration for BitcoinServerMain
[info] org.bitcoins.asyncutil.AsyncUtil$RpcRetryException: Condition timed out after 60 attempts with interval=1 second waiting periods
[info]  at org.bitcoins.tor.config.TorAppConfig.isBinaryFullyStarted(TorAppConfig.scala:166)
[info]  at org.bitcoins.tor.config.TorAppConfig.$anonfun$start$2(TorAppConfig.scala:129)
[info]  at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:434)
[info]  at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:63)
[info]  at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:100)
[info]  at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
[info]  at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:94)
[info]  at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:100)
[info]  at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:49)
[info]  at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:48)
[info]  at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
[info]  at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1016)
[info]  at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1665)
[info]  at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1598)
[info]  at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)

**Review by Christewart — APPROVED** [review]

**Pull request merged**